### DISCUSSION TITLE
How to replace/update a compiled rule's conditions inside knowledge for all future sessions?

### DISCUSSION THREAD

**Discussion by xmxnt** [evrete/evrete#33] [Q&A] [answered]
I have a project where I am adding 1000s of rules to a knowledge instance. The conditions of a rule can change any time. 

I can see there's an example to change conditions of a rule in a session as mentioned [here](https://github.com/evrete/evrete/blob/master/evrete-code-samples/src/main/java/org/evrete/examples/howto/ChangingConditions.java)

I need to know whether there's a way to find a rule by its name and change its condition(LHS) for all future sessions.

**Comment by andbi** [comment, marked as answer]
Yes, it's possible, the `replaceEvaluator` method works for both `Session` and `Knowledge` instances, and in the latter case it will affect all future sessions.

**Reply by xmxnt** [reply]
Thanks a lot. I got it working like this. 

`val newExpression = knowledge.expressionResolver.buildExpression(newConditions, knowledge.getRule(rule.name), knowledge.getRule(rule.name).imports.get(RuleScope.BOTH, RuleScope.LHS) )
                 knowledge.replaceEvaluator(knowledge.get(rule.name + "-handle"),newExpression)`
                 
I am saving  each rule's EvaluatorHandle in knowledge property during knowledge creation. Is there any way I can get the EvaluatorHandle using RuleDescriptor's instance?

**Reply by andbi** [reply]
Great, just remember that the engine can reuse existing conditions. This particularly applies to alpha-conditions, aka predicates over a single type, e.g. `$a.value > 6`.  In that case you will get fewer evaluator handles than the total number of conditions.

**Reply by xmxnt** [reply]
Thank you. Reusing conditions are fine. My understanding is replacing EvaluatorHandle will only delink the selected rule from previous conditions and link it with new ones(or reuse existing) and other rules having same conditions will not be impacted. Correct me if I am wrong.

**Reply by andbi** [reply]
Yes, I just wanted to emphasize that each rule's LHS is composed of condition handles, not conditions, and there's no guarantee that changing a single condition won't affect several rules.

**Reply by xmxnt** [reply]
Aha! Okay. So it's safe to just reload knowledge if there's an update.
